File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/StrategyLine.java
package com.panopset.blackjackEngine;

import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;
import com.panopset.blackjackEngine.Strategy.StratCat;

public class StrategyLine {

  private final BlackjackConfiguration config;
  private final String sourceLine;
  private final StratCat stratCat;
  private final String key;

  public StratCat getStratCat() {
    return stratCat;
  }

  public StrategyLine(StratCat stratCat, String sourceLine, BlackjackConfiguration config) {
    this.stratCat = stratCat;
    this.sourceLine = sourceLine;
    this.config = config;
    key = getStrategyLineElements().get(0);
  }

  public String getSource() {
    return sourceLine;
  }

  public String getAction(final Situation situation) {
    return new StrategyAction(situation,
        new CasinoRules(config.isLateSurrenderAllowed(), config.isDealerHitSoft17(),
            config.getDecks()))
                .getRecommendedAction(getTextForDealerUpCard(situation.dealerUpCard));
  }

  private String getTextForDealerUpCard(BlackjackCard card) {
    return getStrategyLineElements().get(card.getSoftValue() - 1);
  }

  private List<String> strategyLineElements;

  private List<String> getStrategyLineElements() {
    if (strategyLineElements == null) {
      StringTokenizer st = new StringTokenizer(sourceLine, " ");
      strategyLineElements = new ArrayList<>();
      while (st.hasMoreElements()) {
        strategyLineElements.add(st.nextToken());
      }
    }
    return strategyLineElements;
  }

  public String getKey() {
    return key;
  }

  @Override
  public String toString() {
    return getSource();
  }
}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/StrategyLine.kt
package com.panopset.blackjackEngine

import java.util.*

class StrategyLine(val stratCat: StratCat, val source: String, private val config: BlackjackConfiguration) {

    private var strategyLineElements: MutableList<String> = ArrayList()
    init {
        val st = StringTokenizer(source, " ")
        while (st.hasMoreElements()) {
            strategyLineElements.add(st.nextToken())
        }
    }

    val key: String

    fun getAction(situation: Situation): String {
        if (situation.dealerUpCard == null) {
            return CMD_DEAL
        }
        return StrategyAction(
            situation,
            CasinoRules(
                config.isLateSurrenderAllowed,
                config.isDealerHitSoft17,
                config.isDoubleAfterSplitAllowed,
                config.decks
            )
        )
            .getRecommendedAction(getTextForDealerUpCard(situation.dealerUpCard))
    }

    private fun getTextForDealerUpCard(card: BlackjackCard): String {
        return getStrategyLineElements()[card.softValue - 1]
    }


    init {
        key = getStrategyLineElements()[0]
    }

    private fun getStrategyLineElements(): List<String> {
        return strategyLineElements
    }

    override fun toString(): String {
        return source
    }
}
